ngraph/core/shape_inference/include/batch_to_space_shape_inference.hpp

```diff
@@ -0,0 +1,116 @@
+            NODE_VALIDATION_CHECK(op,
+                                  data_rank.get_length() == inputs_same_ps[0].get_length(),
+                                  "block_shape and crop inputs must have same number of elements "
+                                  "as data input rank. Got: ",
+                                  inputs_same_ps[0],
+                                  " and ",
+                                  data_rank);
+        }
+    }
+
+    std::vector<int64_t> block_val, crops_begin_val, crops_end_val;
+
+    if (data_shape.is_static() && get_data_as_int64<T>(1, op, block_val, constant_data) &&
+        get_data_as_int64<T>(2, op, crops_begin_val, constant_data) &&
+        get_data_as_int64<T>(3, op, crops_end_val, constant_data)) {
+        bool block_vals_valid = std::all_of(begin(block_val), end(block_val), [](int64_t elem) {
+            return elem >= 1;
+        });
+        NODE_VALIDATION_CHECK(op, block_vals_valid, "Elements of block_shape input must be greater or equal to one.");
+
+        bool crops_begin_vals_valid = std::all_of(begin(crops_begin_val), end(crops_begin_val), [](int64_t elem) {
+            return elem >= 0;
+        });
+        bool crops_end_vals_valid = std::all_of(begin(crops_end_val), end(crops_end_val), [](int64_t elem) {
+            return elem >= 0;
+        });
+        NODE_VALIDATION_CHECK(op,
+                              crops_begin_vals_valid && crops_end_vals_valid,
+                              "Elements of crops_begin and crops_end inputs must be greater or equal to zero.");
+
+        int64_t block_prod = std::accumulate(begin(block_val), end(block_val), 1, std::multiplies<int64_t>());
+
+        NODE_VALIDATION_CHECK(op,
+                              data_shape[0].get_length() % block_prod == 0,
+                              "The input data's 'batch' axis size: ",
+                              data_shape[0],
+                              " must be a multiple of",
+                              " product of block_shape values: ",
+                              block_prod);
+
+        for (size_t idx = 0; idx < data_shape.size(); idx++) {
+            const bool is_valid_crops_and_shape =
+                crops_begin_val[idx] + crops_end_val[idx] <= block_val[idx] * data_shape[idx].get_length();
+            NODE_VALIDATION_CHECK(op,
+                                  is_valid_crops_and_shape,
+                                  "crops_begin[i] + crops_end[i] must be less or equal to "
+                                  "block_shape[i] * input_shape[i]");
+        }
+
+        auto& output_shape = output_shapes[0];
+        output_shape.resize(data_shape.size());
+        output_shape[0] = data_shape[0].get_length() / block_prod;
+        for (size_t idx = 1; idx < output_shape.size(); ++idx) {
+            output_shape[idx] =
+                data_shape[idx].get_length() * block_val[idx] - crops_begin_val[idx] - crops_end_val[idx];
```

Comment:
this code could be extended for dynamic shape support -- let's assume some dimension of data_shape would be dynamic. For example batch dimension -- we still could propagate spatials correctly